# Laboratorio 3. Deep learning (Ej1 manuscritos)

Andres Emilio Quinto 18288
Andree Toledo 18439

## Instrucciones
Utilice el data set [Digit Recognizer. Learn computer vision fundamentals with the famous MNIST data](https://www.kaggle.com/c/digit-recognizer) de  la  competencia  de Kaggle. Debe  hacer  un  análisis  exploratorio  para  entender  mejor  los datos, sabiendo que el objetivo final es reconocer el carácter que se muestra en la imagen. Recuerde explicar  biencada  uno  de  los  hallazgos  que  haga.La  forma  más  organizada  de  hacer  un  análisis exploratorio  es  generando  ciertas  preguntas  de  las  líneas  que  le  parece  interesante  investigar.Genere un informe con las explicaciones de los pasos que llevó a cabo y los resultados obtenidos. Recuerde  que  la  investigación  debe  ser  reproducible  por  lo  que  debe  guardarel  código  que  ha utilizado  para  resolver  los  ejerciciosy/o  cada  uno  de  los  pasos  llevados  a  cabo  si  utiliza  una herramienta visual. El objetivo es hacer un programa que reconozca caracteres manuscritos.

- - -
### Importación de librerías necesarias para el análisis y algoritmos

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns                       #visualisation
import matplotlib.pyplot as plt             #visualisation

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import tensorflow as tf
import numpy as np
from tensorflow import keras
from sklearn.model_selection import train_test_split

from sklearn.metrics import confusion_matrix
from sklearn.metrics import ConfusionMatrixDisplay
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score

%matplotlib inline     
sns.set(color_codes=True)

- - -
# Análisis exploratorio

Mediante pandas se carga el archivo csv que contiene el [dataset](https://www.kaggle.com/c/digit-recognizer/data).

In [2]:
df = pd.read_csv("train.csv")
# Se muestran las primeras 5 filas
df.head(5)

label  pixel0  pixel1  pixel2  pixel3  pixel4  pixel5  pixel6  pixel7  \
0      1       0       0       0       0       0       0       0       0   
1      0       0       0       0       0       0       0       0       0   
2      1       0       0       0       0       0       0       0       0   
3      4       0       0       0       0       0       0       0       0   
4      0       0       0       0       0       0       0       0       0   

   pixel8  ...  pixel774  pixel775  pixel776  pixel777  pixel778  pixel779  \
0       0  ...         0         0         0         0         0         0   
1       0  ...         0         0         0         0         0         0   
2       0  ...         0         0         0         0         0         0   
3       0  ...         0         0         0         0         0         0   
4       0  ...         0         0         0         0         0         0   

   pixel780  pixel781  pixel782  pixel783  
0         0         0         0         0  
1         0         0         0         0  
2         0         0         0         0  
3         0         0         0         0  
4         0         0         0         0  

[5 rows x 785 columns]

In [3]:
df.tail(5)          
# Se muestran las últimas5 filas

label  pixel0  pixel1  pixel2  pixel3  pixel4  pixel5  pixel6  pixel7  \
41995      0       0       0       0       0       0       0       0       0   
41996      1       0       0       0       0       0       0       0       0   
41997      7       0       0       0       0       0       0       0       0   
41998      6       0       0       0       0       0       0       0       0   
41999      9       0       0       0       0       0       0       0       0   

       pixel8  ...  pixel774  pixel775  pixel776  pixel777  pixel778  \
41995       0  ...         0         0         0         0         0   
41996       0  ...         0         0         0         0         0   
41997       0  ...         0         0         0         0         0   
41998       0  ...         0         0         0         0         0   
41999       0  ...         0         0         0         0         0   

       pixel779  pixel780  pixel781  pixel782  pixel783  
41995         0         0         0         0         0  
41996         0         0         0         0         0  
41997         0         0         0         0         0  
41998         0         0         0         0         0  
41999         0         0         0         0         0  

[5 rows x 785 columns]

In [4]:
df.max()

label        9
pixel0       0
pixel1       0
pixel2       0
pixel3       0
            ..
pixel779    62
pixel780     0
pixel781     0
pixel782     0
pixel783     0
Length: 785, dtype: int64

Se obtienen los tipos de datos de las variables. Se observa que para todas ellas, el tipo es int64. Es decir, todos los datos son números enteros.

In [5]:
df.dtypes

label       int64
pixel0      int64
pixel1      int64
pixel2      int64
pixel3      int64
            ...  
pixel779    int64
pixel780    int64
pixel781    int64
pixel782    int64
pixel783    int64
Length: 785, dtype: object

El dataset cuenta entonces con 785 columnas, siendo la primera de ellas la variable *label*, la cual indica el dígito que fue dibujado por el usuario. Para el resto de columnas, estas representan los valores de los 784 pixeles de una imagen de 28x28. Los valores de los pixeles contienen un número entero que puede ir desde 0 a 255.

En total se tienen cuarenta y dos mil (42,000) filas y 785 columnas.

In [6]:
df.shape

(42000, 785)